### DISCUSSION TITLE
Ask for help to optimize OOM when too many metrics

### DISCUSSION THREAD

**Discussion by yqjdcyy** [apache/skywalking#12513] [Q&A] [answered]
# What work with
Oap:         9.4.0    12* 8C_16G
Agent:      8.14.0
ES:            8* 8C_16G

# What happened

- oap crash
    - trace segment received count from 5K to 10K and 20K
    - oap's **old gen seem to store too much metrics(endpoint_*)** wait for sending to es, because the cpu of es.node 100% all the time
        - informations from dump, and seem to so much endpoint about pool.wapper, like **SwCallableWrapper/pool-251934788-thread-1**
    - keep the water 
        - for 1h then oap will trigger fgc 
        - for longer then oap will be crash

- try something
    1. remove stats for day and hour, less memory
    2. remove no-core metric by oal, less memory and less amount of es.save
    3. server sampling, less es.storage of segments and not work for memory

- the limit of es
    - the cluster of **es is hard to improve** for money
    - try to reduce the amout of metrics maybe the best way

# Ask for Advice
after reading docs and issues, will try to following
1. agent sampling, maybe just work for kafka and consumer but not for metrics and memory
2. agent ignore by `agent.trace.ignore_path`
3. oap group metrics by `endpoint-name-grouping.yml`

**If there are any recommendations or omissions?**

**Comment by wu-sheng** [comment]
I am not sure what you are asking. Elasticsearch scaling usually relies on more nodes, rather than one node.

**Comment by wu-sheng** [comment]
Endpoint number issues are usually caused by restful APIs including parameters. Are you in this case? And why?

Agent should be able to resolve this, as Spring plugins have already known the restful style and parameters from beginning.

Yiu should investigate the reason before asking for simply scaling.

**Reply by yqjdcyy** [reply]
Thanks for your answers.
-----

## more detail
- es.cluster
    - es.cluster with 8 nodes, but it hard to expand so it blocking to fix it immediately.
- endpoint
    - analyze from dump, SwCallableWrapper maybe another reason and blocking in heap to es
        - SwCallableWrapper/pool-251934788-thread-1
    - ![image](https://github.com/user-attachments/assets/7aa93dc1-bf9e-4743-a5ab-0d37e3f795ff)

- top endpoint
    - RESTful API is a good direction, i will try to analyze from `metrics-all`

## for asking
- maybe restful api or swcallable is the reason why the amount so large, how to reduce then?
    - AFAK,`agent.trace.ignore_path` or `endpoint-name-grouping` maybe work, or another?

**Reply by wu-sheng** [reply]
Before answering how, I think you need to measure why.

I can't help until you have use cases.

**Comment by wu-sheng** [comment, marked as answer]
And OAP 9.6+ should have better persistent performance, less reading before flushing metrics back.

Sampling in some case works, but that(agent side sampling) could trade off with metrics accuracy. Server side helps on keeping metrics accuracy and less trace persistence only.

Elasticsearch server should be 1.5x or 2x standards than OAP server.

No matter what we do, again, why matters most.

**Reply by yqjdcyy** [reply]
GET IT！

As you remined me, i loss the GOAL part(work on current configuration of cluster).
And your recommendation about the configuration about oap+es cluster works for me.

Then i will try the oap 9.6+ and analyze the data flow, thx so much.
